internal/receiver/discoveryreceiver/metric_evaluator.go

```diff
@@ -16,26 +16,164 @@ package discoveryreceiver
+	if md.MetricCount() == 0 {
+		return pLogs
+	}
+
+	receiverID, endpointID := statussources.MetricsToReceiverIDs(md)
+	if receiverID == statussources.NoType || endpointID == "" {
+		m.logger.Debug("unable to evaluate metrics from receiver without corresponding name or Endpoint.ID", zap.Any("metrics", md))
+		return pLogs
+	}
+
+	rEntry, ok := m.config.Receivers[receiverID]
+	if !ok {
+		m.logger.Info("No matching configured receiver for metric status evaluation", zap.String("receiver", receiverID.String()))
+		return pLogs
+	}
+	if rEntry.Status == nil || len(rEntry.Status.Metrics) == 0 {
+		return pLogs
+	}
+	var matchFound bool
+
+	stagePLogs := plog.NewLogs()
+	rLog := stagePLogs.ResourceLogs().AppendEmpty()
+	rAttrs := rLog.Resource().Attributes()
+	m.correlateResourceAttributes(
+		md.ResourceMetrics().At(0).Resource().Attributes(), rAttrs,
+		m.correlations.GetOrCreate(receiverID, endpointID),
+	)
+	rAttrs.InsertString(eventTypeAttr, metricMatch)
+	rAttrs.InsertString(receiverRuleAttr, rEntry.Rule)
+
+	logRecords := rLog.ScopeLogs().AppendEmpty().LogRecords()
+
+	receiverMetrics := map[string][]pmetric.Metric{}
+	for i := 0; i < md.ResourceMetrics().Len(); i++ {
+		rm := md.ResourceMetrics().At(i)
+		for j := 0; j < rm.ScopeMetrics().Len(); j++ {
+			sm := rm.ScopeMetrics().At(j)
+			for k := 0; k < sm.Metrics().Len(); k++ {
+				m := sm.Metrics().At(k)
+				receiverMetrics[m.Name()] = append(receiverMetrics[m.Name()], m)
+			}
+		}
+	}
+
+	for status, matches := range rEntry.Status.Metrics {
+		for _, match := range matches {
+			for metricName, metrics := range receiverMetrics {
+				for _, metric := range metrics {
+					if shouldLog, err := m.evaluateMatch(match, metricName, status, receiverID, endpointID); err != nil {
+						m.logger.Info(fmt.Sprintf("Error evaluating %s metric match", status), zap.Error(err))
+						continue
+					} else if !shouldLog {
+						continue
+					}
+					matchFound = true
+					logRecord := logRecords.AppendEmpty()
+					desiredRecord := match.Record
+					if desiredRecord == nil {
+						desiredRecord = &LogRecord{}
+					}
+					var desiredBody string
+					if desiredRecord.Body != "" {
+						desiredBody = desiredRecord.Body
+					}
+					logRecord.Body().SetStringVal(desiredBody)
+					for k, v := range desiredRecord.Attributes {
+						logRecord.Attributes().InsertString(k, v)
+					}
+					severityText := desiredRecord.SeverityText
+					if severityText == "" {
+						severityText = "info"
+					}
+					logRecord.SetSeverityText(severityText)
+					logRecord.Attributes().InsertString(metricNameAttr, metricName)
+					logRecord.Attributes().UpsertString(statusAttr, status)
+					var timestamp pcommon.Timestamp
+					switch metric.DataType() {
+					case pmetric.MetricDataTypeGauge:
+						timestamp = metric.Gauge().DataPoints().At(0).Timestamp()
+					case pmetric.MetricDataTypeSum:
+						timestamp = metric.Sum().DataPoints().At(0).Timestamp()
+					case pmetric.MetricDataTypeHistogram:
```

Comment:
I think you are missing a few metric types: MetricDataTypeExponentialHistogram, MetricDataTypeSummary.
You might need to support MetricDataTypeNone as well. Maybe a default case to log a debug statement would help.

Reply:
Thanks, updated w/ missing types and added more tests for the timestamp extraction.